File: solutions/_983_minimum_cost_for_tickets/solution.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    int mincostTickets(vector<int> &days, vector<int> &costs)
    {
        vector<int> dp(days.back() + 1, 0);
        int i = 0;
        for (int day = 1; day < dp.size(); day++)
        {
            if (day < days[i])
            {
                dp.at(day) = dp.at(day - 1);
            }
            else
            {
                i++;
                dp.at(day) = min({dp.at(day - 1) + costs.at(0),
                                  dp.at(max(0, day - 7)) + costs.at(1),
                                  dp.at(max(0, day - 30)) + costs.at(2)});
            }
        }
        return dp.back();
    }
};

/*
https://leetcode.com/problems/minimum-cost-for-tickets/
*/

File: solutions/_983_minimum_cost_for_tickets/solution.java
package _983_minimum_cost_for_tickets;

import java.util.*;

class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        int[] dp = new int[days[days.length - 1] + 1];
        Set<Integer> set = new HashSet<>();
        for (int day : days) {
            set.add(day);
        }

        for (int i = 1; i < dp.length; i++) {
            if (set.contains(i)) {
                dp[i] = Math.min(
                        dp[i - 1] + costs[0],
                        Math.min(
                                dp[Math.max(0, i - 7)] + costs[1],
                                dp[Math.max(0, i - 30)] + costs[2]));
            } else {
                dp[i] = dp[i - 1];
            }
        }
        return dp[dp.length - 1];
    }
}

/*
 * https://leetcode.com/problems/minimum-cost-for-tickets/
 */